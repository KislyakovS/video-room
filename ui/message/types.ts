import { DetailedHTMLProps, HTMLAttributes } from 'react';

type DefaultProps = DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>;

export enum MessageType {
    my,
    other,
    service,
}

export interface MessageProps extends DefaultProps {
    type: MessageType,
    name?: string;
    text: string;
    fullDate: string;
    time: string;
};