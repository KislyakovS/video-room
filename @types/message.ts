export enum MessageType {
    my,
    other,
    service,
}

export interface Message {
    id: number;
    type: MessageType;
    name: string;
    text: string;
    fullDate: string;
    time: string;
}