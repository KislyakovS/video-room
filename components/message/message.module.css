.my {
    --bubble-background-color: var(--active-color);
    --bubble-text-color: #fff;
}

.other {
    --bubble-background-color: #e6e7ec;
    --bubble-text-color: #333;
}

.message {
    --spacer: 20px;

    display: flex;
}

.my.message {
    flex-direction: row-reverse;
}

.spacer {
    width: var(--spacer);
}

.bubble {
    display: flex;
    flex-direction: column;
    max-width: calc(100% - var(--spacer));
    padding: 5px 8px 5px 10px;

    font-size: 14px;
    color: var(--bubble-text-color);
    overflow-wrap: break-word;
    word-break: break-word;
    white-space: pre-wrap;

    background-color: var(--bubble-background-color);
    border-radius: var(--border-radius);
}

.name {
    font-weight: 700;
}

.text {
    margin: 0;
}

.time {
    font-size: 12px;
}

.my .time {
    margin-left: auto;
}