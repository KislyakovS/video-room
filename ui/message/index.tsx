import { FC } from 'react';
import clsx from 'clsx';

import styles from './message.module.css';

export enum MessageType {
    my,
    other,
    service,
}

interface MessageProps {
    type: MessageType,
    name?: string;
    text: string;
    fullDate: string;
    time: string;
}

const Message: FC<MessageProps> = ({ type, name, text, fullDate, time }) => {
    if (type === MessageType.service) {
        return <p className={styles.serviceMessage}>
            {name && <span className={styles.name}>{name} </span>}
            {text}
        </p>
    }

    const messageClasses = clsx(styles.message, {
        [styles.my]: type === MessageType.my,
        [styles.other]: type === MessageType.other
    });

    return (
        <div className={messageClasses}>
            <div className={styles.bubble}>
                {name && <span className={styles.name}>{name}</span>}
                <p className={styles.text}>{text}</p>
                <time className={styles.time} dateTime={fullDate}>{time}</time>
            </div>
            <span className={styles.spacer} />
        </div>
    );
};

export default Message;