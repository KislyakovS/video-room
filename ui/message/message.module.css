.my {
    --bubble-background-color: var(--active-color);
    --bubble-text-color: var(--second-color);
}

.other {
    --bubble-background-color: #e6e7ec;
    --bubble-text-color: var(--primary-color);
}

.message {
    --spacer: 20px;

    display: flex;
}

.serviceMessage {
    margin: 0;
    padding: 4px 0;

    text-align: center;
}

.my.message {
    flex-direction: row-reverse;
}

.spacer {
    width: var(--spacer);
}

.bubble {
    display: flex;
    flex-direction: column;
    max-width: calc(100% - var(--spacer));
    padding: 5px 8px 5px 10px;

    font-size: 14px;
    color: var(--bubble-text-color);
    overflow-wrap: break-word;
    word-break: break-word;
    white-space: pre-wrap;

    background-color: var(--bubble-background-color);
    border-radius: var(--border-radius);
}

.name {
    font-weight: 700;
}

.serviceMessage .name {
    color: var(--active-color);
}

.text {
    margin: 0;
}

.time {
    font-size: 12px;
}

.my .time {
    margin-left: auto;
}
