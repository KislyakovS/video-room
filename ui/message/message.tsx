import { FC } from 'react';
import clsx from 'clsx';

import MessageProps, { MessageType } from './message.type';

import styles from './message.module.css';

const Message: FC<MessageProps> = ({ className, type, name, text, fullDate, time, ...props }) => {
    if (type === MessageType.service) {
        return <div {...props} className={clsx(className, styles.serviceMessage)}>
            {name && <span className={styles.name}>{name} </span>}
            {text}
        </div>
    }

    const hasTime = fullDate && time;
    const messageClasses = clsx(className, styles.message, {
        [styles.my]: type === MessageType.my,
        [styles.other]: type === MessageType.other
    });

    return (
        <div {...props} className={messageClasses}>
            <div className={styles.bubble}>
                {name && <span className={styles.name}>{name}</span>}
                <p className={styles.text}>{text}</p>
                {hasTime && <time className={styles.time} dateTime={fullDate}>{time}</time>}
            </div>
            <span className={styles.spacer} />
        </div>
    );
};

export default Message;