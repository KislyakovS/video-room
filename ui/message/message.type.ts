import { DetailedHTMLProps, HTMLAttributes } from 'react';

import { MessageType } from '@types';

type DefaultProps = DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>;

export default interface MessageProps extends DefaultProps {
    type: MessageType,
    name?: string;
    text: string;
    fullDate?: string;
    time?: string;
};