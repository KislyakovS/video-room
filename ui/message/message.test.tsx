import { render } from '@testing-library/react';

import Message from './message';
import { MessageType } from './message.type';

describe('Message Component:', () => {
    it('snapshot type my', () => {
        const { container } = render(<Message
            type={MessageType.my}
            text='Hello!'
            fullDate='2022-11-14 12:20'
            time='12:20'
        />);

        expect(container).toMatchSnapshot();
    });

    it('snapshot type other', () => {
        const { container } = render(<Message
            type={MessageType.other}
            text='Hello!'
            name='Fyodor Dostoevsky'
            fullDate='2022-11-14 12:20'
            time='12:20'
        />);

        expect(container).toMatchSnapshot();
    });

    it('snapshot type service', () => {
        const { container } = render(<Message
            type={MessageType.service}
            text='left the room.'
            name='Fyodor Dostoevsky'
        />);

        expect(container).toMatchSnapshot();
    });
});