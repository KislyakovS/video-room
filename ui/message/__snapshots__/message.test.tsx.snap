// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message Component: snapshot type my 1`] = `
<div>
  <div
    class="message my"
  >
    <div
      class="bubble"
    >
      <p
        class="text"
      >
        Hello!
      </p>
      <time
        class="time"
        datetime="2022-11-14 12:20"
      >
        12:20
      </time>
    </div>
    <span
      class="spacer"
    />
  </div>
</div>
`;

exports[`Message Component: snapshot type other 1`] = `
<div>
  <div
    class="message other"
  >
    <div
      class="bubble"
    >
      <span
        class="name"
      >
        Fyodor Dostoevsky
      </span>
      <p
        class="text"
      >
        Hello!
      </p>
      <time
        class="time"
        datetime="2022-11-14 12:20"
      >
        12:20
      </time>
    </div>
    <span
      class="spacer"
    />
  </div>
</div>
`;

exports[`Message Component: snapshot type service 1`] = `
<div>
  <div
    class="serviceMessage"
  >
    <span
      class="name"
    >
      Fyodor Dostoevsky
       
    </span>
    left the room.
  </div>
</div>
`;
